coreax/coreset.py

```diff
@@ -0,0 +1,118 @@
+    of the coreset :math:`\nu_\hat{n}` has (approximately for some algorithms) the same
+    "centre-of-mass" as the push-forward measure for the original data :math:`\mu_n`
+
+    .. math:
+        \text{CoM}(\mu_n) = \text{CoM}(\nu_\hat{n}),
+        \text{CoM}(\nu_\hat{n}) = \int_\Omega \Phi(\omega) d\nu_\hat{x}(\omega),
+        \text{CoM}(\nu_\hat{n}) = \sum_{i=1}^\hat{n} \hat{w}_i \delta_{\Phi(\hat{x}_i)}.
+
+    .. note:
+        Depending on the algorithm, the test-functions may be explicitly specified by
+        the user, or implicitly defined by the algorithm's specific objectives.
+
+    :param nodes: The (weighted) coreset nodes, math:`x_i \in \text{supp}(\nu_\hat{n})`;
+        once instantiated, the nodes should be accessed via :meth:`Coresubset.coreset`
+    :param pre_coreset_data: The dataset :math:`X` used to construct the coreset.
+    """
+
+    nodes: Data = eqx.field(converter=_convert_to_weighted)
+    pre_coreset_data: _Data
+
+    def __len__(self):
+        """Return Coreset size/length."""
+        return len(self.nodes)
+
+    @property
+    def coreset(self) -> Data:
+        """Materialised coreset."""
+        return self.nodes
+
+
+class Coresubset(Coreset[_Data], Generic[_Data]):
+    r"""
+    Data structure for representing a coresubset.
+
+    A coresubset is a :class`Coreset`, with the additional condition that the support of
+    the reduced measure (the coreset), must be a subset of the support of the original
+    measure (the original data), such that
+
+    .. math:
+        \hat{x}_i = x_i, \forall i \in I,
+        I \subset \{1, \dots, n\}, text{card}(I) = \hat{n}.
+
+    Thus, a coresubset, unlike a corset, ensures that feasibility constraints on the
+    support of the measure are maintained :cite:`litterer2012recombination`. This is
+    vital if, for example, the test-functions are only defined on the support of the
+    original measure/nodes, rather than all of :math:`\Omega`.
+
+    In coresubsets, the measure reduction can be implicit (setting weights/nodes to
+    zero for all :math:`i \in I \ {1, \dots, n}`) or explicit (removing entries from the
+    weight/node arrays). The implicit approach is useful when input/output array shape
+    stability is required (E.G. for some JAX transformations); the explicit approach is
+    more similar to a standard coreset.
+
+    :param nodes: The (weighted) coresubset node indices, :math:`I`; the materialised
+        coresubset nodes should be accessed via :meth:`Coresubset.coreset`.
+    :param pre_coreset_data: The dataset :math:`X` used to construct the coreset.
+    """
+
+    # Incompatibility between Pylint and eqx.field. Pyright handles this correctly.
+    # pylint: disable=no-member
+    @property
+    def coreset(self) -> Data:
+        """Materialise the coresubset from the indices and original data."""
+        coreset_nodes = self.pre_coreset_data.data[self.unweighted_indices]
```

Comment:
Maybe I am misunderstanding, but if this is a supervised Coresubset then we won't be materialising the supervision here right?

Reply:
This was by design. My logic was that the coreset should only be the weighted nodes (this is consistent with the docstrings). If you require the supervision associated with each coreset node/index, then you can get this with `Coreset.pre_coreset_data.supervision[self.unweighted_indices]`, providing the `pre_coreset_data` is `SupervisedData`.

Reply:
I see, makes sense!